src/core/algorithms/nar/des/encoded_nar.cpp

```diff
@@ -0,0 +1,102 @@
+#include "encoded_nar.h"
+
+#include "algorithms/nar/value_range.h"
+#include "model/types/types.h"
+
+namespace algos::des {
+
+using model::NAR;
+
+size_t EncodedNAR::VectorSize() {
+    return encoded_value_ranges_.size() * EncodedValueRange().kFieldCount + 1;
+}
+
+size_t EncodedNAR::FeatureCount() {
+    return encoded_value_ranges_.size();
+}
+
+// TODO: remove code duplication here
+double& EncodedNAR::operator[](size_t index) {
+    qualities_consistent_ = false;
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
+
+double const& EncodedNAR::operator[](size_t index) const {
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
```

Comment:
code duplication, extract insides to a private method, smth like:
```cpp
double& EncodedNAR::operator[](size_t index) {
    qualities_consistent_ = false;
    return GetElementAtIndex(index);
}

double const& EncodedNAR::operator[](size_t index) const {
    return GetElementAtIndex(index);
}
```

Reply:
`GetElementAtIndex(index)` then has to be a const function returning a non-const reference. This won't compile:

```
double& EncodedNAR::GetElementAtIndex(size_t index) const {
    if (index == 0) {
        return implication_sign_pos_;
    } else {
        index--;
        size_t feature = index / EncodedValueRange().kFieldCount;
        size_t feature_field = index % EncodedValueRange().kFieldCount;
        return encoded_value_ranges_[feature][feature_field];
    }
}
```
because "qualifiers dropped in binding reference of type "double &" to initializer of type "const double". I'm not sure how to fix this.

Reply:
I believe smth like that will work:
```cpp
template <typename EncodedNARType>
auto& GetElementAtIndex(EncodedNARType& self, size_t index) {
    //body
}
```

and call like `return GetElementAtIndex(*this, index);`

Reply:
If I understood correctly we then have to reference `self.implication_sign_pos` and `implication_sign_pos` is private and inaccessible in that function 😭. GetElementAtIndex must be a private method of EncodedNAR. But then we have to either make it const, making it unable to return a non-const ref, or not const, making it unusable in a const method. I guess a macro would work.

https://stackoverflow.com/questions/15413831/c-two-versions-of-overloading-subscript-operator

one of the commenters said:

> Yes, you'll need to define both of them. You could define the non-const version in terms of the const version but then you'd have to do a const_cast. You're better off just having the duplicate implementation.

Reply:
Okay, makes sense, thanks